presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -0,0 +1,603 @@
+        }
+
+        private static final class SubfieldExtractor
+                extends DefaultExpressionTraversalVisitor<Void, Context>
+        {
+            private final TypeProvider typeProvider;
+
+            private SubfieldExtractor(TypeProvider typeProvider)
+            {
+                this.typeProvider = requireNonNull(typeProvider, "typeProvider is null");
+            }
+
+            @Override
+            protected Void visitSubscriptExpression(SubscriptExpression node, Context context)
+            {
+                Optional<Subfield> subfield = toSubfield(node);
+                if (subfield.isPresent()) {
+                    context.subfields.add(subfield.get());
+                }
+                else {
+                    process(node.getBase(), context);
+                    process(node.getIndex(), context);
+                }
+                return null;
+            }
+
+            @Override
+            protected Void visitDereferenceExpression(DereferenceExpression node, Context context)
+            {
+                Optional<Subfield> subfield = toSubfield(node);
+                if (subfield.isPresent()) {
+                    context.subfields.add(subfield.get());
+                }
+                else {
+                    process(node.getBase(), context);
+                    process(node.getField(), context);
+                }
+                return null;
+            }
+
+            @Override
+            protected Void visitSymbolReference(SymbolReference node, Context context)
+            {
+                context.variables.add(new VariableReferenceExpression(node.getName(), typeProvider.get(Symbol.from(node))));
+                return null;
+            }
+        }
+
+        private static Optional<Subfield> toSubfield(Node expression)
+        {
+            ImmutableList.Builder<Subfield.PathElement> elements = ImmutableList.builder();
+            while (true) {
+                if (expression instanceof SymbolReference) {
+                    return Optional.of(new Subfield(((SymbolReference) expression).getName(), elements.build().reverse()));
+                }
+
+                if (expression instanceof DereferenceExpression) {
+                    DereferenceExpression dereference = (DereferenceExpression) expression;
+                    elements.add(new NestedField(dereference.getField().getValue()));
+                    expression = dereference.getBase();
+                }
+                else if (expression instanceof SubscriptExpression) {
+                    SubscriptExpression subscript = (SubscriptExpression) expression;
+                    Expression index = subscript.getIndex();
+                    if (index instanceof Cast) {
+                        index = ((Cast) index).getExpression();
+                    }
+                    if (index instanceof LongLiteral) {
```

Comment:
nit: does it make sense to use `else if` here ?

Reply:
No, because we need to process the expression inside of cast.